### PR TITLE
Sparkmaxes -> core

### PR DESCRIPTION
**Pull request by greenblitz4590**
all the sparkmax motor with all things

### CODE REVIEW

File: src/main/java/frc/robot/hardware/motor/cansparkmax/SparkMaxMotor.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import frc.robot.hardware.ConnectedInputAutoLogged;
+import frc.robot.hardware.motor.IMotor;
+import frc.robot.hardware.signal.InputSignal;
+import frc.robot.hardware.signal.cansparkmax.ISparkMaxSignal;
+import frc.utils.alerts.Alert;
+import frc.utils.alerts.AlertManager;
+import frc.utils.alerts.PeriodicAlert;
+import org.littletonrobotics.junction.Logger;
+
+
```
Comment (drorta) [inline src/main/java/frc/robot/hardware/motor/cansparkmax/SparkMaxMotor.java:14]: ❤️

File: src/main/java/frc/robot/hardware/motor/phoenix6/TalonFXMotor.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+		NeutralModeValue neutralModeValue = brake ? NeutralModeValue.Brake : NeutralModeValue.Coast;
+		motor.setNeutralMode(neutralModeValue);
+	}
+
+	@Override
+	public void resetPosition(Rotation2d position) {
+		motor.setPosition(position.getRotations());
+	}
+
+
```
Comment (drorta) [inline src/main/java/frc/robot/hardware/motor/phoenix6/TalonFXMotor.java:41]: ❤️

File: src/main/java/frc/robot/hardware/motor/phoenix6/TalonFXMotor.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	public void stop() {
+		motor.stopMotor();
+	}
+
+	@Override
+	public void setPower(double power) {
+		motor.set(power);
+	}
+
+
```
Comment (drorta) [inline src/main/java/frc/robot/hardware/motor/phoenix6/TalonFXMotor.java:52]: ❤️

**Review by drorta — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/hardware/motor/cansparkmax/BrushedSparkMAXMotor.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package frc.robot.hardware.motor.cansparkmax;
+
+import com.revrobotics.CANSparkLowLevel;
+import com.revrobotics.CANSparkMax;
+
+public class BrushedSparkMAXMotor extends SparkMaxMotor {
+
+	protected final CANSparkMax motor;
```
Comment (greenblitz4590) [inline src/main/java/frc/robot/hardware/motor/cansparkmax/BrushedSparkMAXMotor.java:8]: private

File: src/main/java/frc/robot/hardware/motor/cansparkmax/SparkMaxMotor.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	}
+
+	@Override
+	public void updateSignals(InputSignal... signals) {
+		for (InputSignal signal : signals) {
+			if(signal instanceof ISparkMaxSignal){
+				Logger.processInputs(logPath, signal);
+			}
+		}
+		connectedInput.connected = motor.getBusVoltage() > 0;
```
Comment (NoamRosenberg08) [inline src/main/java/frc/robot/hardware/motor/cansparkmax/SparkMaxMotor.java:54]: add logger.process input(...,connected)

**Review by NoamRosenberg08 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxDoubleSignal.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
+package frc.robot.hardware.signal.cansparkmax;
+
+import frc.robot.hardware.signal.DoubleSignal;
+
+import java.util.function.Supplier;
+
+public class SparkMaxDoubleSignal extends DoubleSignal implements ISparkMaxSignal {
+    Supplier<Double> doubleSignalSupplier;
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxDoubleSignal.java:8]: private final

File: src/main/java/frc/robot/hardware/motor/cansparkmax/SparkMaxMotor.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+import frc.utils.alerts.Alert;
+import frc.utils.alerts.AlertManager;
+import frc.utils.alerts.PeriodicAlert;
+import org.littletonrobotics.junction.Logger;
+
+
+public abstract class SparkMaxMotor implements IMotor {
+
+	protected final CANSparkMax motor;
+	protected final String logPath;
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/motor/cansparkmax/SparkMaxMotor.java:18]: privatw

File: src/main/java/frc/robot/hardware/motor/cansparkmax/SparkMaxMotor.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+	private final ConnectedInputAutoLogged connectedInput;
+
+	public SparkMaxMotor(CANSparkMax motor, String logPath) {
+		this.motor = motor;
+		this.logPath = logPath;
+
+		connectedInput = new ConnectedInputAutoLogged();
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/motor/cansparkmax/SparkMaxMotor.java:25]: this.

File: src/main/java/frc/robot/hardware/motor/cansparkmax/SparkMaxMotor.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+import org.littletonrobotics.junction.Logger;
+
+
+public abstract class SparkMaxMotor implements IMotor {
+
+	protected final CANSparkMax motor;
+	protected final String logPath;
+	private final ConnectedInputAutoLogged connectedInput;
+
+	public SparkMaxMotor(CANSparkMax motor, String logPath) {
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/motor/cansparkmax/SparkMaxMotor.java:21]: logpath first

**Review by hillelv — DISMISSED** [review]

File: src/main/java/frc/robot/hardware/motor/cansparkmax/SparkMaxMotor.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+	private final ConnectedInputAutoLogged connectedInput;
+
+	public SparkMaxMotor(String logPath, CANSparkMax motor) {
+		this.logPath = logPath;
+		this.motor = motor;
+
+		this.connectedInput = new ConnectedInputAutoLogged();
+		connectedInput.connected = true;
+
+		AlertManager.addAlert(new PeriodicAlert(Alert.AlertType.WARNING, logPath + "disconnectedAt", () -> !isConnected()));
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/hardware/motor/cansparkmax/SparkMaxMotor.java:27]: why is this a warning

**Review by bossyshrimp68 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/hardware/motor/phoenix6/TalonFXMotor.java [hunk 11]
```diff
@@ -38,7 +38,6 @@ public void resetPosition(Rotation2d position) {
 		motor.setPosition(position.getRotations());
 	}
 
-
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/motor/phoenix6/TalonFXMotor.java:41]: return this

File: src/main/java/frc/robot/hardware/motor/phoenix6/TalonFXMotor.java [hunk 12]
```diff
@@ -49,7 +48,6 @@ public void setPower(double power) {
 		motor.set(power);
 	}
 
-
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/motor/phoenix6/TalonFXMotor.java:52]: this tooo

File: src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxAngleSignal.java [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+	private final Supplier<Double> angleSignalSupplier;
+
+	public SparkMaxAngleSignal(String name, Supplier<Double> angleSignalSupplier, AngleUnit angleUnit) {
+		super(name, angleUnit);
+		this.angleSignalSupplier = angleSignalSupplier;
+	}
+
+	@Override
+	protected double getNewValue() {
+		return angleSignalSupplier.get();
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxAngleSignal.java:19]: add timestamp

File: src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxDoubleSignal.java [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+	private final Supplier<Double> doubleSignalSupplier;
+
+	public SparkMaxDoubleSignal(String name, Supplier<Double> doubleSignalSupplier) {
+		super(name);
+		this.doubleSignalSupplier = doubleSignalSupplier;
+	}
+
+	@Override
+	protected Double getNewValue() {
+		return doubleSignalSupplier.get();
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxDoubleSignal.java:18]: here toooo

File: src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxDoubleSignal.java [hunk 15]
```diff
@@ -0,0 +1,21 @@
+package frc.robot.hardware.signal.cansparkmax;
+
+import frc.robot.hardware.signal.DoubleSignal;
+
+import java.util.function.Supplier;
+
+public class SparkMaxDoubleSignal extends DoubleSignal implements ISparkMaxSignal {
+
+	private final Supplier<Double> doubleSignalSupplier;
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxDoubleSignal.java:9]:
```suggestion
	private final Supplier<Double> valueSupplier;
```

File: src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxAngleSignal.java [hunk 16]
```diff
@@ -0,0 +1,22 @@
+package frc.robot.hardware.signal.cansparkmax;
+
+import frc.robot.hardware.signal.AngleSignal;
+import frc.utils.AngleUnit;
+
+import java.util.function.Supplier;
+
+public class SparkMaxAngleSignal extends AngleSignal implements ISparkMaxSignal {
+
+	private final Supplier<Double> angleSignalSupplier;
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxAngleSignal.java:10]:
```suggestion
	private final Supplier<Double> angleSupplier;
```

File: src/main/java/frc/robot/hardware/motor/cansparkmax/BrushedSparkMAXMotor.java [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+import com.revrobotics.CANSparkMax;
+
+public class BrushedSparkMAXMotor extends SparkMaxMotor {
+
+	private final CANSparkMax motor;
+
+	public BrushedSparkMAXMotor(String logPath, CANSparkMax motor) {
+		super(logPath, motor);
+		if (motor.getMotorType() != CANSparkLowLevel.MotorType.kBrushed) {
+			throw new IllegalArgumentException("inserted brushless sparkmax motor to brushed!");
```
[the remark this answers is not in the document]
Comment (YHerm) [reply src/main/java/frc/robot/hardware/motor/cansparkmax/BrushedSparkMAXMotor.java:13]:
```suggestion
			throw new IllegalArgumentException("inserted BrushlessSparkMAXMotor  to BrushedSparkMAXMotor!");
```

File: src/main/java/frc/robot/hardware/motor/cansparkmax/BrushlessSparkMAXMotor.java [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+import frc.robot.hardware.request.cansparkmax.SparkMaxAngleRequest;
+import frc.robot.hardware.request.cansparkmax.SparkMaxDoubleRequest;
+import frc.utils.calibration.sysid.SysIdCalibrator;
+
+import java.util.function.BiFunction;
+
+public class BrushlessSparkMAXMotor extends SparkMaxMotor implements ControllableMotor {
+
+	private final CANSparkMax motor;
+	private final BiFunction<Rotation2d, Rotation2d, Rotation2d> feedforward;
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/motor/cansparkmax/BrushlessSparkMAXMotor.java:18]: remove ff

File: src/main/java/frc/robot/hardware/motor/cansparkmax/BrushlessSparkMAXMotor.java [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+
+	public BrushlessSparkMAXMotor(
+		CANSparkMax motor,
+		BiFunction<Rotation2d, Rotation2d, Rotation2d> feedforward,
+		SysIdRoutine.Config sysidConfig,
+		String logPath
+	) {
+		super(logPath, motor);
+		if (motor.getMotorType() != CANSparkLowLevel.MotorType.kBrushless) {
+			throw new IllegalArgumentException("inserted brushed sparkmax motor to brushless!");
```
[the remark this answers is not in the document]
Comment (YHerm) [reply src/main/java/frc/robot/hardware/motor/cansparkmax/BrushlessSparkMAXMotor.java:29]: like in the the before

File: src/main/java/frc/robot/hardware/motor/cansparkmax/SparkMaxMotor.java [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+			}
+		}
+		connectedInput.connected = motor.getBusVoltage() > 0;
+		Logger.processInputs(logPath, connectedInput);
+	}
+
+	@Override
+	public boolean isConnected() {
+		return connectedInput.connected;
+	}
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/motor/cansparkmax/SparkMaxMotor.java:60]: put this above the other overrides

**Review by YHerm — CHANGES_REQUESTED** [review]
do gradlew SpotlessAPPLY

File: src/main/java/frc/robot/hardware/motor/cansparkmax/BrushlessSparkMAXMotor.java
Comment (YHerm) [inline src/main/java/frc/robot/hardware/motor/cansparkmax/BrushlessSparkMAXMotor.java:1]: Move spark max wrapper and device is to here

Comment (YHerm) [inline src/main/java/frc/robot/hardware/motor/cansparkmax/BrushlessSparkMAXMotor.java:1]: I think the pack name can be just spark or rev

File: src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxDoubleSignal.java [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+	private final Supplier<Double> valueSupplier;
+
+	public SparkMaxDoubleSignal(String name, Supplier<Double> doubleSignalSupplier) {
+		super(name);
+		this.valueSupplier = doubleSignalSupplier;
+	}
+
+	@Override
+	protected TimedValue<Double> getNewValue() {
+		return new TimedValue<>(valueSupplier.get(), System.currentTimeMillis());
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxDoubleSignal.java:19]: Why system and not timer or Halutil or logger?

Comment (YHerm) [reply src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxDoubleSignal.java:19]: If the talonfx return like system keep it

File: src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxAngleSignal.java [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+	private final Supplier<Double> valueSupplier;
+
+	public SparkMaxAngleSignal(String name, Supplier<Double> angleSignalSupplier, AngleUnit angleUnit) {
+		super(name, angleUnit);
+		this.valueSupplier = angleSignalSupplier;
+	}
+
+	@Override
+	protected TimedValue<Double> getNewValue() {
+		return new TimedValue<>(valueSupplier.get(), System.currentTimeMillis());
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxAngleSignal.java:20]: Here too

File: src/main/java/frc/robot/hardware/motor/sparkmax/BrushedSparkMAXMotor.java [hunk 23]
```diff
@@ -0,0 +1,18 @@
...
+import com.revrobotics.CANSparkMax;
+
+public class BrushedSparkMAXMotor extends SparkMaxMotor {
+
+	private final CANSparkMax motor;
+
+	public BrushedSparkMAXMotor(String logPath, CANSparkMax motor) {
+		super(logPath, motor);
+		if (motor.getMotorType() != CANSparkLowLevel.MotorType.kBrushed) {
+			throw new IllegalArgumentException("\"inserted BrushlessSparkMAXMotor to BrushedSparkMAXMotor!\"");
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/motor/sparkmax/BrushedSparkMAXMotor.java:13]: Why the “/“

File: src/main/java/frc/robot/hardware/motor/sparkmax/BrushlessSparkMAXMotor.java [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+
+public class BrushlessSparkMAXMotor extends SparkMaxMotor implements ControllableMotor {
+
+	private final CANSparkMax motor;
+	private final SysIdCalibrator.SysIdConfigInfo sysIdConfigInfo;
+
+	public BrushlessSparkMAXMotor(CANSparkMax motor, SysIdRoutine.Config sysidConfig, String logPath) {
+		super(logPath, motor);
+		if (motor.getMotorType() != CANSparkLowLevel.MotorType.kBrushless) {
+			throw new IllegalArgumentException("\"inserted BrushedSparkMAXMotor to BrushlessSparkMAXMotor!\"");
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/motor/sparkmax/BrushlessSparkMAXMotor.java:22]: Why the “/“

File: src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxAngleSignal.java [hunk 25]
```diff
@@ -0,0 +1,24 @@
...
+import frc.robot.hardware.signal.AngleSignal;
+import frc.robot.hardware.signal.TimedValue;
+import frc.utils.AngleUnit;
+
+import java.util.function.Supplier;
+
+public class SparkMaxAngleSignal extends AngleSignal implements ISparkMaxSignal {
+
+	private final Supplier<Double> valueSupplier;
+
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxAngleSignal.java:13]: angle supplier

File: src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxDoubleSignal.java [hunk 26]
```diff
@@ -0,0 +1,23 @@
...
+	private final Supplier<Double> valueSupplier;
+
+	public SparkMaxDoubleSignal(String name, Supplier<Double> doubleSignalSupplier) {
+		super(name);
+		this.valueSupplier = doubleSignalSupplier;
+	}
+
+	@Override
+	protected TimedValue<Double> getNewValue() {
+		return new TimedValue<>(valueSupplier.get(), HALUtil.getHALRuntimeType());
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxDoubleSignal.java:20]: Runtime type?

File: src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxDoubleSignal.java [hunk 27]
```diff
@@ -0,0 +1,23 @@
...
+import frc.robot.hardware.signal.DoubleSignal;
+import frc.robot.hardware.signal.TimedValue;
+
+import java.util.function.Supplier;
+
+public class SparkMaxDoubleSignal extends DoubleSignal implements ISparkMaxSignal {
+
+	private final Supplier<Double> valueSupplier;
+
+	public SparkMaxDoubleSignal(String name, Supplier<Double> doubleSignalSupplier) {
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxDoubleSignal.java:13]:
```suggestion
	public SparkMaxDoubleSignal(String name, Supplier<Double> valueSupplier) {
```

File: src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxAngleSignal.java [hunk 28]
```diff
@@ -0,0 +1,24 @@
...
+	public SparkMaxAngleSignal(String name, Supplier<Double> angleSignalSupplier, AngleUnit angleUnit) {
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxAngleSignal.java:14]:
```suggestion
	public SparkMaxAngleSignal(String name, Supplier<Double> angleSupplier, AngleUnit angleUnit) {
```

File: src/main/java/frc/robot/hardware/motor/sparkmax/SparkMaxMotor.java [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+import frc.robot.hardware.signal.InputSignal;
+import frc.robot.hardware.signal.cansparkmax.ISparkMaxSignal;
+import frc.utils.alerts.Alert;
+import frc.utils.alerts.AlertManager;
+import frc.utils.alerts.PeriodicAlert;
+import org.littletonrobotics.junction.Logger;
+
+public abstract class SparkMaxMotor implements IMotor {
+
+	private final CANSparkMax motor;
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/motor/sparkmax/SparkMaxMotor.java:16]: Make protected instead of 2 pointers

**Review by YHerm — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxAngleSignal.java [hunk 30]
```diff
@@ -0,0 +1,24 @@
...
+		super(name, angleUnit);
+		this.valueSupplier = angleSignalSupplier;
+	}
+
+	@Override
+	protected TimedValue<Double> getNewValue() {
+		return new TimedValue<>(valueSupplier.get(), HALUtil.getHALRuntimeType());
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/signal/cansparkmax/SparkMaxAngleSignal.java:21]: runtime type?

**Review by bossyshrimp68 — DISMISSED** [review]

**Pull request merged**